New endpoint add custom habit

LiliaMokhnatska opened the pull request:
## Summary Of Issue :
Need to add new endpoint for adding custom habit

**Issue Link**
https://github.com/ita-social-projects/GreenCity/issues/5309

## Summary Of Changes :
1) added fields  isCustomHabit (by default is false) and userId (not required)
2) added change log for these values 
3) added method addCustomHabit in controller  HabitController  @PostMapping("/custom/{userId}")
4) added  "/habit/custom/{userId}" in SecurityConfig 
5)  added method  addCustomHabit  in HabitService
6) implemented addCustomHabit in HabitServiceImp
7) added method findTagsByNames in TagsRepo
8) added  class AddCustomHabitDtoRequest 
9) added class AddCustomHabitDtoResponse
10) added class CustomHabitMapper
11) added class CustomShoppingListMapper 
12) added class HabitTranslationMapper 
13)added tests for methods in controller, service and mappers 


## CHECK LIST
- [x] Code is up-to-date with the `dev` branch.
- [x] You've successfully built and run the tests locally.
- [x] There are new or updated unit tests validating the change.
- [ ] JIRA/ Github Issue number & title in PR title (ISSUE-: Ticket title)
- [x] This template filled (above this section).
- [x] Sonar's report does not contain bugs, vulnerabilities, security issues, code smells ar duplication
- [ ] `NEED_REVIEW` and `READY_FOR_REVIEW` labels are added.
- [x] All files reviewed before sending to reviewers

LiliaMokhnatska committed:
added endpoint for creating custom habit, changed dtos and mappers, added tests, added new endpoint in SecurityConfig, added default value in habits changelog, changed documentation of method in TagsRepo,
changed params in method in HabitController

RomaMocherniuk commented on core/src/main/java/greencity/controller/HabitController.java:
```diff
@@ -133,4 +137,30 @@ public ResponseEntity<PageableDto<HabitDto>> getAllByTagsAndLanguageCode(
+    })
+    @PostMapping("/custom/{userId}")
+    public ResponseEntity<AddCustomHabitDtoResponse> addCustomHabit(
+        @PathVariable Long userId,
```
Can every user create a habit for some other user (when `loggedUserId != userId`)?

LiliaMokhnatska replied:
I changed it. Now, we can't

RomaMocherniuk commented on dao/src/main/java/greencity/entity/Habit.java:
```diff
@@ -48,6 +48,12 @@ public class Habit {
     private Integer defaultDuration;
 
+    @Column(name = "is_custom_habit", nullable = false)
+    private Boolean isCustomHabit;
```
Do we need it?
I do not see where we use it

LiliaMokhnatska replied:
This field is set for the future so that we can later check if the custom habit.

RomaMocherniuk commented on dao/src/main/java/greencity/entity/Habit.java:
```diff
@@ -48,6 +48,12 @@ public class Habit {
+    @Column(name = "is_custom_habit", nullable = false)
+    private Boolean isCustomHabit;
+
+    @Column(name = "user_id")
```
can `user_id` be `null`?

LiliaMokhnatska replied:
Yes, these fields will not be null for the custom habit in order to know who create  custom habit, but null for the default one.

RomaMocherniuk commented on service-api/src/main/java/greencity/dto/habit/AddCustomHabitDtoRequest.java:
```diff
@@ -0,0 +1,28 @@
+@EqualsAndHashCode
+public class AddCustomHabitDtoRequest {
+    @Min(value = 1, message = ServiceValidationConstants.HABIT_COMPLEXITY)
+    @Max(value = 3, message = ServiceValidationConstants.HABIT_COMPLEXITY)
```
please add test for validation

LiliaMokhnatska replied:
added

RomaMocherniuk commented on service-api/src/main/java/greencity/dto/habit/AddCustomHabitDtoResponse.java:
```diff
@@ -0,0 +1,29 @@
+    private Long userId;
+    @Min(value = 1, message = ServiceValidationConstants.HABIT_COMPLEXITY)
+    @Max(value = 3, message = ServiceValidationConstants.HABIT_COMPLEXITY)
+    private String image;
```
it seems like this validation is not related to the `image`

LiliaMokhnatska replied:
Yes, it should be for the "complexity" field. There is a mistake.

RomaMocherniuk replied:
If you had tests for validation it would be easy to catch it
It is a good example of why tests are important

LiliaMokhnatska replied:
added

RomaMocherniuk commented on service/src/main/java/greencity/mapping/CustomHabitMapper.java:
```diff
@@ -0,0 +1,19 @@
+            .image(addCustomHabitDtoRequest.getImage())
+            .complexity(addCustomHabitDtoRequest.getComplexity())
+            .defaultDuration(addCustomHabitDtoRequest.getDefaultDuration())
+            .isCustomHabit(true)
```
Should we add `user_id` too for custom habit?

LiliaMokhnatska replied:
Yes, we set it in method in Service

RomaMocherniuk replied:
Why cant we set `user_id` here?

LiliaMokhnatska replied:
I changed parameters in method in controller. Now we find the user from the principals and in order to set userId we use UserRepo userRepo.findByEmail(userEmail)

RomaMocherniuk commented on service/src/main/java/greencity/service/HabitServiceImpl.java:
```diff
@@ -2,17 +2,24 @@
+import greencity.mapping.CustomHabitMapper;
+import greencity.mapping.CustomShoppingListMapper;
+import greencity.mapping.HabitTranslationMapper;
+import greencity.repository.*;
```
please do not use `*`

LiliaMokhnatska replied:
changed

RomaMocherniuk commented on service/src/main/java/greencity/service/HabitServiceImpl.java:
```diff
@@ -126,4 +141,33 @@ public List<Long> addAllShoppingListItemsByListOfId(Long habitId, List<Long> lis
     }
+
+    @Override
+    public AddCustomHabitDtoResponse addCustomHabit(Long userId,
```
should this method be annotated `@Transactional`?

LiliaMokhnatska replied:
changed

RomaMocherniuk commented on service/src/test/java/greencity/ModelUtils.java:
```diff
@@ -196,12 +197,7 @@
-import java.util.HashSet;
-import java.util.List;
-import java.util.Set;
+import java.util.*;
```
please do not use `*`

LiliaMokhnatska replied:
changed

RomaMocherniuk commented on service/src/test/java/greencity/ModelUtils.java:
```diff
@@ -2294,4 +2290,110 @@ public static UserShoppingAndCustomShoppingListsDto getUserShoppingAndCustomShop
             .build();
     }
+
+    public static AddCustomHabitDtoRequest getAddCustomHabitDtoRequest() {
```
please move values you use at least twice to the constants

LiliaMokhnatska replied:
changed

RomaMocherniuk commented on service/src/main/java/greencity/service/HabitServiceImpl.java:
```diff
@@ -126,4 +141,33 @@ public List<Long> addAllShoppingListItemsByListOfId(Long habitId, List<Long> lis
+        for (String languageCode : languageCodes) {
+            habitTranslationList.forEach(habitTranslation -> habitTranslation.setLanguage(languageRepo
+                .findByCode(languageCode)
+                .orElseThrow(NoSuchElementException::new)));
```
please add tests for this case

LiliaMokhnatska replied:
added test for this case.

LiliaMokhnatska committed:
deleted unused imports

RomaMocherniuk commented on core/src/test/java/greencity/ModelUtils.java:
```diff
@@ -109,10 +111,7 @@
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.List;
+import java.util.*;
```
please do not use `*`

LiliaMokhnatska replied:
changed

RomaMocherniuk commented on service-api/src/main/java/greencity/service/HabitService.java:
```diff
@@ -83,4 +85,15 @@ public interface HabitService {
+     * Method to save {@link AddCustomHabitDtoResponse}.
+     *
+     * @param userEmail                {@link String} - user email.
+     * @param addCustomHabitDtoRequest dto with {@link AddCustomHabitDtoRequest}
```
please use the same order as you use in the method  `addCustomHabitDtoRequest addCustomHabitDtoRequest, String userEmail`

LiliaMokhnatska replied:
changed

RomaMocherniuk commented on service-api/src/test/java/greencity/dto/habit/AddCustomHabitDtoRequestTest.java:
```diff
@@ -0,0 +1,65 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class AddCustomHabitDtoRequestTest {
```
please add tests for cases when `complexity` is `null` and lower than `min`

LiliaMokhnatska replied:
added

RomaMocherniuk commented on service-api/src/test/java/greencity/dto/habit/AddCustomHabitDtoResponseTest.java:
```diff
@@ -0,0 +1,65 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class AddCustomHabitDtoResponseTest {
```
same here

LiliaMokhnatska replied:
added

RomaMocherniuk commented on service/src/test/java/greencity/mapping/CustomHabitMapperTest.java:
```diff
@@ -0,0 +1,30 @@
+@ExtendWith(MockitoExtension.class)
+class CustomHabitMapperTest {
+    @InjectMocks
+    CustomHabitMapper customHabitMapper;
```
can be `private`

LiliaMokhnatska replied:
changed

RomaMocherniuk commented on service/src/test/java/greencity/mapping/CustomShoppingListMapperTests.java:
```diff
@@ -0,0 +1,48 @@
+@ExtendWith(MockitoExtension.class)
+class CustomShoppingListMapperTests {
+    @InjectMocks
+    CustomShoppingListMapper customShoppingListMapper;
```
`private`

LiliaMokhnatska replied:
changed

LiliaMokhnatska committed:
added tests for values of complexity null and less than min value
in AddCustomHabitDtoResponseTest and AddCustomHabitDtoRequestTest

changed access modifier for  @InjectMocks in CustomHabitMapperTest, CustomHabitMapperTest, HabitTranslationMapperTests

changed documentation for method addCustomHabit in HabitService

added constants in ModelUtils

LiliaMokhnatska committed:
changed imports
